### ISSUE TITLE
sst deploy fails after Cognito user pool schema changes are wrongly detected

### ISSUE THREAD

**Issue by paulswail** [anomalyco/sst#3928]
I've hit a blocker issue with `sst deploy` and `sst diff` detecting that my Cognito user pool's schema field has changed and it then attempts to deploy it and fails with the error "cannot modify or remove schema items: provider=aws@6.56.1".
This error would be expected given that Cognito doesn't allow this, however, these fields have not changed in my code or in my AWS account and had been working fine through several deploys until I ran `sst refresh` when attempting to resolve a separate issue (Neon provider plugin crashed).  From then onwards, I get this error when I run deploy.

I'm using SST v3.2.36.

Here's my user pool definition:

```ts
export const mainUserPool = new sst.aws.CognitoUserPool("MainPool", {
  aliases: ["email"],
  transform: {
    userPool: {
      schemas: [
        {
          name: "user_id",
          attributeDataType: "String",
          developerOnlyAttribute: false,
          mutable: true,
          required: false,
        },
      ],
    },
  },
});

```

When I run `sst diff`, I get: 

```
*  MainPool sst:aws:CognitoUserPool → MainPoolUserPool aws:cognito:UserPool
   * schemas[0].attributeDataType = String
   + schemas[0].developerOnlyAttribute = false
   * schemas[0].mutable = true
   * schemas[0].name = user_id
   + schemas[0].required = false
   * schemas[0].stringAttributeConstraints
```

**Comment by paulswail** [comment]
Would be nice if the `sst diff` command had a `--verbose` argument to show for `*` items that have changed, what was the value it's being changed from so I could see why it thinks this field has changed

**Comment by jayair** [comment]
That's strange that it would detect that something changed when there were no changes. Could it be possible that the AWS provider was updated in between? Or did this happen between deploys without any changes in between?

**Comment by paulswail** [comment]
Re: AWS provider bring updated, I didn’t explicitly run sst install. Is there another way this could have been updated?

This first happened yesterday and I ended up just deleting the whole stack to get unblocked. I redeployed new stack today and it had been deploying fine through multiple deploys until I ran sst refresh.

Referencing this Discord thread where another user hit the same issue: https://discord.com/channels/983865673656705025/983865673656705028/1290605674656694296

**Comment by paulswail** [comment]
Adding some more info.
Running `sst state edit` to open the local state file, this is what is in the `outputs` of my user pool resource: 

```json
"schemas": [
    {
        "attributeDataType": "String",
        "developerOnlyAttribute": false,
        "mutable": true,
        "name": "user_id",
        "numberAttributeConstraints": null,
        "required": false,
        "stringAttributeConstraints": {
            "maxLength": null,
            "minLength": null
        }
    }
],
```

And when I use `aws cognito-idp describe-user-pool --user-pool-id XXXX --query 'UserPool.SchemaAttributes'`, this is what I get:

```json
{
        "Name": "custom:user_id",
        "AttributeDataType": "String",
        "DeveloperOnlyAttribute": false,
        "Mutable": true,
        "Required": false,
        "StringAttributeConstraints": {}
    }
```

I can't spot anything that the diff command would flag as having changed.

**Comment by arjunyel** [comment]
I'm getting this issue as well except ignoreChanges is not letting me work around. Did you end up finding a fix?

**Comment by brunocleite** [comment]
> That's strange that it would detect that something changed when there were no changes. Could it be possible that the AWS provider was updated in between? Or did this happen between deploys without any changes in between?

I am experiencing the same issue. 
Did a clean deploy, a refresh and then faced the issue on a new deploy. 
No updates on AWS Provider in between.

**Comment by arjunyel** [comment]
Example of how to ignore cognito changes

```
transform: {
    userPool: (args, opts) => {
      opts.ignoreChanges = ["schemas"]; // uncomment this to workaround this issue: https://github.com/sst/ion/issues/1319
      args.schemas = [
        {
          name: "user_id",
          attributeDataType: "String",
          developerOnlyAttribute: false,
          mutable: true,
          required: false,
        },
      ];
      args.passwordPolicy = { // redacted};
    },
  },
```

**Comment by jayair** [comment]
@fwang let's take a look at this?

**Comment by fwang** [comment]
@brunocleite did setting `opts.ignoreChanges = ["schemas"];` suggested by @arjunyel work for you?

**Comment by brunocleite** [comment]
> @brunocleite did setting `opts.ignoreChanges = ["schemas"];` suggested by @arjunyel work for you?

Yes. The suggested workaround works. @fwang

**Comment by imdkbj** [comment]
Hi guys, Can we have an update on this.
I have tried the ignoreChanges but not working

<img width="516" alt="Image" src="https://github.com/user-attachments/assets/13e7fbd4-f8e1-4c94-a409-9d689c9fd642" />

**Comment by paulswail** [comment]
> Hi guys, Can we have an update on this. I have tried the ignoreChanges but not working
> 
> <img alt="Image" width="516" src="https://private-user-images.githubusercontent.com/46839821/408049365-13e7fbd4-f8e1-4c94-a409-9d689c9fd642.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgyMjk0NTcsIm5iZiI6MTczODIyOTE1NywicGF0aCI6Ii80NjgzOTgyMS80MDgwNDkzNjUtMTNlN2ZiZDQtZjhlMS00Yzk0LWE0MDktOWQ2ODljOWZkNjQyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTMwVDA5MjU1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5NDQwOTI3ZTc5NDkzNDRiMWVmMzBhYjYyZTFmMDFjMGNhZWE0OWI2MzZjMWFlODQ5MjdmNDhmNTQyYTMzYmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VBOYZ3nfGNlhxC2n5uNUFITxybMrQmnxRhmTkx3PcZ8">

@imdkbj I was able to get the ignoreChanges to work when I changed the transform definition from an object format (like you have it) to a function, as per https://github.com/sst/sst/issues/3928#issuecomment-2563805547

**Comment by imdkbj** [comment]
Hey @paulswail , Thanks for the quick reply.

I tried this way too

<img width="555" alt="Image" src="https://github.com/user-attachments/assets/b9e47cca-25a7-48dd-8cab-34a241d328ff" />

Looks like it is trying to create a new userpool here
```
|  Error       UserPoolexternalUserPool aws:cognito:UserPool
|  unable to replace resource "urn:pulumi:dev::sso::sst:aws:CognitoUserPool$aws:cognito/userPool:UserPool::UserPoolexternalUserPool"
as it is currently marked for protection. To unprotect the resource, remove the `protect` flag from the resource in your Pulumi program and run `pulumi up`
```

**Comment by paulswail** [comment]
> Hey [@paulswail](https://github.com/paulswail) , Thanks for the quick reply.
> 
> I tried this way too
> 
> <img alt="Image" width="555" src="https://private-user-images.githubusercontent.com/46839821/408076561-b9e47cca-25a7-48dd-8cab-34a241d328ff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgyMzU5MzMsIm5iZiI6MTczODIzNTYzMywicGF0aCI6Ii80NjgzOTgyMS80MDgwNzY1NjEtYjllNDdjY2EtMjVhNy00OGRkLThjYWItMzRhMjQxZDMyOGZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTMwVDExMTM1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiYjM4MTFkNTllY2Y0OWY4NmMwMGUxOTlkNGFkMDEyYjIyYjkwY2ZmOTE5NGZkYzcwOTFhYTJiN2FhZmYyM2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jp5GtKi_jkO6FMPfbb2Mz0Pyz95FjzYUqMc9JX2fDxs">
> Looks like it is trying to create a new userpool here
> 
> ```
> |  Error       UserPoolexternalUserPool aws:cognito:UserPool
> |  unable to replace resource "urn:pulumi:dev::sso::sst:aws:CognitoUserPool$aws:cognito/userPool:UserPool::UserPoolexternalUserPool"
> as it is currently marked for protection. To unprotect the resource, remove the `protect` flag from the resource in your Pulumi program and run `pulumi up`
> ```

Hmm not totally sure if this is the issue, but I see you are re-assigning the `args` parameter object rather than modifying. So perhaps the changes in your newly declared `args` object are not being applied? Could you try explicitly assigning the individual fields of `args`, e.g. `args.passwordPolicy = passwordPolicy`, etc ?

**Comment by imdkbj** [comment]
It's worked via this way

Object.assign(args, configs.userPoolConfig)

But have to delete and re-create groups, This might be due to i have changed group lately. Have to think how to handle in prod.

**Comment by mike-van** [comment]
> > Hi guys, Can we have an update on this. I have tried the ignoreChanges but not working
> > <img alt="Image" width="516" src="https://private-user-images.githubusercontent.com/46839821/408049365-13e7fbd4-f8e1-4c94-a409-9d689c9fd642.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgyMjk0NTcsIm5iZiI6MTczODIyOTE1NywicGF0aCI6Ii80NjgzOTgyMS80MDgwNDkzNjUtMTNlN2ZiZDQtZjhlMS00Yzk0LWE0MDktOWQ2ODljOWZkNjQyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTMwVDA5MjU1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5NDQwOTI3ZTc5NDkzNDRiMWVmMzBhYjYyZTFmMDFjMGNhZWE0OWI2MzZjMWFlODQ5MjdmNDhmNTQyYTMzYmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VBOYZ3nfGNlhxC2n5uNUFITxybMrQmnxRhmTkx3PcZ8">
> 
> [@imdkbj](https://github.com/imdkbj) I was able to get the ignoreChanges to work when I changed the transform definition from an object format (like you have it) to a function, as per [#3928 (comment)](https://github.com/sst/sst/issues/3928#issuecomment-2563805547)

this works for me, by changing from object to a function based `transform.userPool`, the ignoreChanges finally works. weirdly for my case, it shows a new undeclared attribute in my schema in my sst diff output:
```
schemas = [
       ...my other attributes,
       {
         "attributeDataType": null,
         "developerOnlyAttribute": null,
         "mutable": null,
         "name": null,
         "numberAttributeConstraints": null,
         "required": null,
         "stringAttributeConstraints": null
       }
     ]
```

**Comment by js-maniac** [comment]
@mike-van I'm also getting the same when running sst diff
